rfd/0098-securing-docker-images.md

```diff
@@ -0,0 +1,181 @@
+---
+authors: Trent Clarke (trent@goteleport.com)
+state: draft
+---
+
+# RFD 0098 - Delivering secure Teleport OCI Images
+
+# Required Approvers
+* Engineering: @r0mant
+* Security: @reed
+* Product: (@xinding33 || @klizhentas)
+
+## What
+
+This RFD discusses structures and processes to increase the security of the OCI 
+images we use to deliver Teleport to our customers.
+
+## Why
+
+One of our shipping channels is a Docker image. As the provider of that docker
+image we are (at least partialy) responsible for everything on it, not just
+Teleport. We should not ship vulnerabilities to our clients, even if those 
+vulnerabilities are not odirectly in our software.
+
+## Details
+
+### Goals
+
+For the sake of this RFD, I will define _delivering a secure image_ as 
+
+> Reliably producing an OCI image that:
+>
+>
+>  1. a priori is unlikely to contain vulnerabilities in and of itself, 
+>  2. contains software of known provenance,
+>  3. has no warnings or vulnerabilities flagged when run through a reputable 
+>     scanner at the time of creation, and
+>  4. flags vulnerabilities even after creation, either in Teleport itself or 
+>     a dependency.
+
+### Non-goals
+
+This RFD will not discuss building images for and/or running Teleport on
+platforms other than Linux.
+
+### Approach
+
+This RFD describes a 2-pronged approach for meeting the above goals:
+
+   1. Switching to distroless base images, and 
+   2. Automated, ongoing monitoring of images using a scanning service
+
+### 1. Distroless Images
+
+Distroless images contain only an application and the minimal set of
```

Comment:
Do all of our features work in distroless images? I'm thinking things like PAM, BPF, Rust (GLIBC dependency), etc.

Reply:
We don't enable SSH service in any of our Helm charts so I'm not sure getting things like PAM and BPF working is needed tbh. Not sure if any of our customers enable SSH service in their custom configs, or what a use-case would even be since we have Kube access.

Couple of items that we might need to think about though are:

- Desktop access is also currently not present at all in Helm charts. Do we need/want to support it? If so, we might need to make sure Rust stuff works, yeah.
- We will likely be looking at adding BPF-like advanced session recording to Kubernetes access relatively soon. No idea how it would work yet but I wonder if moving to distroless now is going to complicate things somewhat.

Reply:
I've picked the CC image specifically for GLIBC support, which is all our Rust code should need. 

As far as PAM and BPF support is concerned, my understanding is that we only nee specific kernel versions to make this work, and we link `libpam` etc statically. It's definitely worth some practical testing, though.

Reply:
And practical tests confirm that I'm wrong, at least for PAM. Looking into adding PAM support

Reply:
We might be able to go back to the distroless source and rebuild our image from there - after all, if _they_ can add arbitrary debian packages to their images then we should be able to as well. 

I'm trying to understand `bazel` enough to see if this is a reasonable option. (The default GHA builder image has `bazel` already installed, so that's a start)

Reply:
Answer: You _can_ add arbitrary Debian packages to an image using `bazel`, but the method used by `rules_docker` and `distroless` is functionally no different from the unpack-and-copy method used in the Dockerfile, so there doesn;t seem to be much point in making the tooling more complicated.